File: src/main/java/ch/obermuhlner/astro/gradient/operation/SubtractSplineImageOperation.java
package ch.obermuhlner.astro.gradient.operation;

import ch.obermuhlner.astro.gradient.math.SplineInterpolator;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class SubtractSplineImageOperation extends AbstractSimpleChannelImageOperation {

  private final double factor;
  private final double[] xyPairs;

  public SubtractSplineImageOperation(double factor, double... xyPairs) {
    this.factor = factor;
    this.xyPairs = xyPairs;
  }

  @Override
  protected double channelOperation(double channel1, double channel2, int x, int y, int channelIndex) {
    if (channel2 <= 0.0) {
      return channel1;
    }

    List<Double> xPoints = new ArrayList<>();
    List<Double> yPoints = new ArrayList<>();

    xPoints.add(0.0);
    yPoints.add(0.0);

    xPoints.add(channel2);
    yPoints.add(channel2 * factor);

    for (int i = 0; i < xyPairs.length; i+=2) {
      xPoints.add(xyPairs[i + 0]);
      yPoints.add(xyPairs[i + 1]);
    }

    xPoints.add(1.0);
    yPoints.add(1.0);

    SplineInterpolator spline = SplineInterpolator.createMonotoneCubicSpline(
        xPoints,
        yPoints
    );

    return spline.interpolate(channel1);
  }

  @Override
  public String toString() {
    return "SubtractSpine(factor=" + factor + ", xy=" + Arrays.toString(xyPairs) + ")";
  }
}

File: src/main/java/ch/obermuhlner/astro/gradient/operation/SubtractSplineImageOperation.kt
package ch.obermuhlner.astro.gradient.operation

import ch.obermuhlner.astro.gradient.math.SplineInterpolator
import java.util.*

class SubtractSplineImageOperation constructor(private val factor: Double, private vararg val xyPairs: Double) : AbstractSimpleChannelImageOperation() {
    override fun channelOperation(channel1: Double, channel2: Double, x: Int, y: Int, channelIndex: Int): Double {
        if (channel2 <= 0.0) {
            return channel1
        }
        val xPoints: MutableList<Double> = ArrayList()
        val yPoints: MutableList<Double> = ArrayList()
        xPoints.add(0.0)
        yPoints.add(0.0)
        xPoints.add(channel2)
        yPoints.add(channel2 * factor)
        var i: Int = 0
        while (i < xyPairs.size) {
            xPoints.add(xyPairs[i + 0])
            yPoints.add(xyPairs[i + 1])
            i += 2
        }
        xPoints.add(1.0)
        yPoints.add(1.0)
        val spline: SplineInterpolator = SplineInterpolator.createMonotoneCubicSpline(
                xPoints,
                yPoints
        )
        return spline.interpolate(channel1)
    }

    override fun toString(): String {
        return "SubtractSpine(factor=" + factor + ", xy=" + xyPairs.contentToString() + ")"
    }
}